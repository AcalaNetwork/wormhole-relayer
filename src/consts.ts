import {
  CHAIN_ID_ACALA,
  CHAIN_ID_KARURA,
} from '@certusone/wormhole-sdk';
import { ROUTER_TOKEN_INFO } from '@acala-network/asset-router/dist/consts';
import dotenv from 'dotenv';

dotenv.config({ path: '.env' });

const isTestnet = Number(process.env.TESTNET_MODE ?? 1);

export const enum ETH_RPC {
  LOCAL = 'http://localhost:8545',

  BSC = 'https://endpoints.omniatech.io/v1/bsc/mainnet/public',
  KARURA = 'https://eth-rpc-karura.aca-api.network',
  ACALA = 'https://eth-rpc-acala.aca-api.network',

  BSC_TESTNET = 'https://endpoints.omniatech.io/v1/bsc/testnet/public',
  FUJI = 'https://avalanche-fuji-c-chain.publicnode.com',
  GOERLI = 'https://rpc.ankr.com/eth_goerli',
  KARURA_TESTNET = 'https://eth-rpc-karura-testnet.aca-staging.network',
};

export const enum BSC_TOKEN {
  USDC = '0xB04906e95AB5D797aDA81508115611fee694c2b3',
  DAI = '0x3413a030EF81a3dD5a302F4B4D11d911e12ed337',
  USDT = '0x337610d27c682e347c9cd60bd4b3b107c9d34ddd',
};

export const enum FUJI_TOKEN {
  USDC = '0x63A30f239DC8d1c17Bf6653a68Fc6C2F83641E6d',
};

export const WORMHOLE_GUARDIAN_RPC = {
  TESTNET: ['https://wormhole-v2-testnet-api.certus.one'],
  MAINNET: ['https://wormhole-v2-mainnet-api.certus.one'],
};

const RELAYER_BASE_URL = 'http://localhost:3111';
// const RELAYER_BASE_URL = 'https://relayer.aca-dev.network';
// const RELAYER_BASE_URL = 'https://relayer.aca-api.network';

export const RELAYER_API = {
  SHOULD_RELAY: '/shouldRelay',
  RELAY: '/relay',

  SHOULD_ROUTE_XCM: '/shouldRouteXcm',
  ROUTE_XCM: '/routeXcm',

  SHOULD_ROUTE_WORMHOLE: '/shouldRouteWormhole',
  ROUTE_WORMHOLE: '/routeWormhole',
  RELAY_AND_ROUTE: '/relayAndRoute',
  RELAY_AND_ROUTE_BATCH: '/relayAndRouteBatch',

  GET_HOMA_ROUTER_ADDR: '/shouldRouteHoma',
  ROUTE_HOMA: '/routeHoma',

  NO_ROUTE: '/noRoute',
  VERSION: '/version',
  TEST_TIMEOUT: '/testTimeout',
  HEALTH: '/health',
} as const;

export const RELAYER_URL = {
  SHOULD_RELAY: `${RELAYER_BASE_URL}${RELAYER_API.SHOULD_RELAY}`,
  RELAY: `${RELAYER_BASE_URL}${RELAYER_API.RELAY}`,

  SHOULD_ROUTE_XCM: `${RELAYER_BASE_URL}${RELAYER_API.SHOULD_ROUTE_XCM}`,
  ROUTE_XCM: `${RELAYER_BASE_URL}${RELAYER_API.ROUTE_XCM}`,

  SHOULD_ROUTE_WORMHOLE: `${RELAYER_BASE_URL}${RELAYER_API.SHOULD_ROUTE_WORMHOLE}`,
  ROUTE_WORMHOLE: `${RELAYER_BASE_URL}${RELAYER_API.ROUTE_WORMHOLE}`,
  RELAY_AND_ROUTE: `${RELAYER_BASE_URL}${RELAYER_API.RELAY_AND_ROUTE}`,
  RELAY_AND_ROUTE_BATCH: `${RELAYER_BASE_URL}${RELAYER_API.RELAY_AND_ROUTE_BATCH}`,

  GET_HOMA_ROUTER_ADDR: `${RELAYER_BASE_URL}${RELAYER_API.GET_HOMA_ROUTER_ADDR}`,
  ROUTE_HOMA: `${RELAYER_BASE_URL}${RELAYER_API.ROUTE_HOMA}`,

  NO_ROUTE: `${RELAYER_BASE_URL}${RELAYER_API.NO_ROUTE}`,
  VERSION: `${RELAYER_BASE_URL}${RELAYER_API.VERSION}`,
  TEST_TIMEOUT: `${RELAYER_BASE_URL}${RELAYER_API.TEST_TIMEOUT}`,
  HEALTH: `${RELAYER_BASE_URL}${RELAYER_API.HEALTH}`,
} as const;

/* ---------------
   thredhold amount is defined as "amount that will show on VAA"
   address should be **lower case** address to be consistent
                                                --------------- */
const RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS_DEV = {
  [CHAIN_ID_KARURA]: {
    // 0.1 BSC USDT => karura WUSDT 0x478dEFc2Fc2be13a505dafBDF1e5400847E2efF6
    '0x337610d27c682e347c9cd60bd4b3b107c9d34ddd': '10000000',

    // 0.02 BSC ETH => karura WETH ?
    '0xd66c6b4f0be8ce5b39d52e0fd1344c389929b378': '2000000',

    // 0.001 Goerli USDC => karura WUSDC 0xE5BA1e8E6BBbdC8BbC72A58d68E74B13FcD6e4c7
    '0x07865c6e87b9f70255377e024ace6630c1eaa37f': '1000',
  },
  [CHAIN_ID_ACALA]: {
    // 0.1 BSC USDT => Acala WUSDT 0xb54bA7F042DCAFba0A546e69F69E81b4F59B9C92
    '0x337610d27c682e347c9cd60bd4b3b107c9d34ddd': '10000000',

    // 0.02 BSC ETH => karura WETH ?
    '0xd66c6b4f0be8ce5b39d52e0fd1344c389929b378': '2000000',

    // 0.01 Goerli USDC => karura WUSDC 0xE5BA1e8E6BBbdC8BbC72A58d68E74B13FcD6e4c7
    '0x07865c6e87b9f70255377e024ace6630c1eaa37f': '10000',
  },
};

const RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS_PROD = {
  [CHAIN_ID_KARURA]: {
    // 10 ETH mainnet USDC => karura WUSDC 0x1F3a10587A20114EA25Ba1b388EE2dD4A337ce27
    '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48': '10000000',     // 1 USDC = 10{6}

    // 10 ETH mainnet DAI  => karura WDAI 0x4bB6afB5Fa2b07a5D1c499E1c3DDb5A15e709a71
    '0x6b175474e89094c44da98b954eedeac495271d0f': '1000000000',   // 1 DAI = 10{8}

    // 1 Acala AUSD        => karura WAUSD 0xE20683ad1ED8bbeED7E1aE74Be10F19D8045B530
    '0x0000000000000000000100000000000000000001': '100000000',    // 1 AUSD = 10{8}
  },
  [CHAIN_ID_ACALA]: {
    // 10 ETH mainnet USDC => karura WUSDC 0x07DF96D1341A7d16Ba1AD431E2c847d978BC2bCe
    '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48': '10000000',     // 1 USDC = 10{6}

    // 10 ETH mainnet DAI  => karura WDAI 0x54A37A01cD75B616D63E0ab665bFfdb0143c52AE
    '0x6b175474e89094c44da98b954eedeac495271d0f': '1000000000',   // 1 DAI = 10{8}

    // 1 karura WAUSD      => Acala AUSD 0x0000000000000000000100000000000000000001
    '0x0000000000000000000100000000000000000001': '100000000',   // 1 WAUSD = 10{8}
  },
};

export const RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS = isTestnet
  ? RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS_DEV
  : RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS_PROD;

export const enum PARA_ID {
  BASILISK = '2090',
  CALAMARI = '2084',
  SHADOW = '2012',

  HYDRA = '2034',
  MANTA = '2104',
  CRUST =  '2008',
}

export const GOERLI_USDC = '0x07865c6e87b9f70255377e024ace6630c1eaa37f';
export const ZERO_ADDR = '0x0000000000000000000000000000000000000000';

const ROUTE_SUPPORTED_CHAINS_AND_ASSETS_DEV = {
  //   [HYDRA_PARA_ID]: [
  //     GOERLI_USDC,
  //   ],
  [PARA_ID.BASILISK]: [
    GOERLI_USDC,
  ],
} as const;

const ASSETS_COMMON_HYDRA = [
  ROUTER_TOKEN_INFO.weth.originAddr,
  ROUTER_TOKEN_INFO.wbtc.originAddr,
] as const;

const ASSETS_COMMON_MANTA = [
  ROUTER_TOKEN_INFO.arb.originAddr,
  ROUTER_TOKEN_INFO.ldo.originAddr,
  ROUTER_TOKEN_INFO.shib.originAddr,
  ROUTER_TOKEN_INFO.wmatic.originAddr,
  ROUTER_TOKEN_INFO.wbnb.originAddr,
  ROUTER_TOKEN_INFO.uni.originAddr,
  ROUTER_TOKEN_INFO.busd.originAddr,
  ROUTER_TOKEN_INFO.link.originAddr,
  ROUTER_TOKEN_INFO.ape.originAddr,
] as const;

export const ROUTE_SUPPORTED_CHAINS_AND_ASSETS_PROD = {
  /* --------------- karura --------------- */
  [PARA_ID.BASILISK]: [
    ...ASSETS_COMMON_HYDRA,
    ROUTER_TOKEN_INFO.usdc.originAddr,
  ],
  [PARA_ID.CALAMARI]: [
    ...ASSETS_COMMON_MANTA,
    ROUTER_TOKEN_INFO.wbtc.originAddr,
    ROUTER_TOKEN_INFO.weth.originAddr,
    ROUTER_TOKEN_INFO.usdc.originAddr,
    ROUTER_TOKEN_INFO.usdt.originAddr,
  ],
  [PARA_ID.SHADOW]: [
    ROUTER_TOKEN_INFO.csm.originAddr,
  ],

  /* --------------- acala --------------- */
  [PARA_ID.HYDRA]: [
    ...ASSETS_COMMON_HYDRA,
    ROUTER_TOKEN_INFO.dai.originAddr,
    ROUTER_TOKEN_INFO.ape.originAddr,
  ],
  [PARA_ID.MANTA]: [
    ...ASSETS_COMMON_MANTA,
  ],
  [PARA_ID.CRUST]: [
    ROUTER_TOKEN_INFO.cru.originAddr,
  ],
} as const;

export const ROUTE_SUPPORTED_CHAINS_AND_ASSETS = isTestnet
  ? ROUTE_SUPPORTED_CHAINS_AND_ASSETS_DEV
  : ROUTE_SUPPORTED_CHAINS_AND_ASSETS_PROD;

export const DEST_PARA_ID_TO_ROUTER_WORMHOLE_CHAIN_ID = {
  [PARA_ID.BASILISK]: CHAIN_ID_KARURA,
  [PARA_ID.CALAMARI]: CHAIN_ID_KARURA,
  [PARA_ID.SHADOW]: CHAIN_ID_KARURA,

  [PARA_ID.HYDRA]: CHAIN_ID_ACALA,
  [PARA_ID.MANTA]: CHAIN_ID_ACALA,
  [PARA_ID.CRUST]: CHAIN_ID_ACALA,
} as const;

export const ROUTER_TOKEN_INFO_TESTNET = {
  usdc: {
    originChain: 'ETH',
    originAddr: GOERLI_USDC,
    karuraAddr: '0xE5BA1e8E6BBbdC8BbC72A58d68E74B13FcD6e4c7',
    acalaAddr: null,
    decimals: 6,
    fee: 0.04,
  },
} as const;

export const TESTNET_MODE_WARNING = `
  ----------------------------
  üî® running in testnet mode
  ‚ùå don't use it for mainnet!
  ----------------------------
`;

export const VERSION = '1.5.1';
