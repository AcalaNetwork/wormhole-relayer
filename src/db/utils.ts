import { isEvmAddress } from '@acala-network/eth-providers';
import { prisma } from './client';

export interface RouterInfo {
  params: string;       // stringified JSON params
  factoryAddr: string;
  recipient: string;
  feeAddr: string;
  routerAddr: string;
}

export const connectDb = async () => {
  try {
    await prisma.$connect();
  } catch (err) {
    console.error('❗️ db connection failed!');
    throw err;
  }
};

const _sanitizeParams = (params: string) => {
  let jsonParams: Record<string, string> = {};
  try {
    jsonParams = JSON.parse(params);
  } catch (err) {
    return params;
  }

  Object.keys(jsonParams).forEach(key => {
    if (isEvmAddress(jsonParams[key])) {
      jsonParams[key] = jsonParams[key].toLowerCase();
    }
  });

  return JSON.stringify(jsonParams);
};

// use lowercase for all addrs to be consistent
export const sanitizeRouterInfo = <T extends Partial<RouterInfo>>(info: T): T => {
  return {
    ...(info.params && { params: _sanitizeParams(info.params) }),
    ...(info.factoryAddr && { factoryAddr: info.factoryAddr.toLowerCase() }),
    ...(info.recipient && { recipient: info.recipient.toLowerCase() }),
    ...(info.feeAddr && { feeAddr: info.feeAddr.toLowerCase() }),
    ...(info.routerAddr && { routerAddr: info.routerAddr.toLowerCase() }),
  } as T;
};
