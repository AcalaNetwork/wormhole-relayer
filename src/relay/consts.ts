import {
  CHAIN_ID_KARURA,
  CHAIN_ID_ACALA,
} from '@certusone/wormhole-sdk';

// export const RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS = {
//   [CHAIN_ID_KARURA]: {
//     '0xddb64fe46a91d46ee29420539fc25fd07c5fea3e': '10000000',   // 0.1 WETH
//     '0x2d8be6bf0baa74e0a907016679cae9190e80dd0a': '10000000',   // 0.1 ERC20
//   },
//   [CHAIN_ID_ACALA]: {
//     '0x12345': '10000',
//   },
// };

// thredhold amount is defined as "amount that will show on VAA"
// address should be lower case address to be consistent
export const RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS = {
  [CHAIN_ID_KARURA]: {
    // 0.1 BSC USDT => karura WUSDT 0x478dEFc2Fc2be13a505dafBDF1e5400847E2efF6
    '0x337610d27c682e347c9cd60bd4b3b107c9d34ddd': '10000000',

    // 0.02 BSC ETH => karura WETH ?
    '0xd66c6b4f0be8ce5b39d52e0fd1344c389929b378': '2000000',
  },
  [CHAIN_ID_ACALA]: {
    // 0.1 BSC USDT => Acala WUSDT 0xb54bA7F042DCAFba0A546e69F69E81b4F59B9C92
    '0x337610d27c682e347c9cd60bd4b3b107c9d34ddd': '10000000',

    // 0.02 BSC ETH => karura WETH ?
    '0xd66c6b4f0be8ce5b39d52e0fd1344c389929b378': '2000000',
  },
};
