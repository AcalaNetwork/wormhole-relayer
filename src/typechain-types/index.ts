/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type * as lib from "./lib";
export type { lib };
import type * as src from "./src";
export type { src };
export * as factories from "./factories";
export type { IERC20 } from "./lib/openzeppelin-contracts/contracts/token/ERC20/IERC20";
export { IERC20__factory } from "./factories/lib/openzeppelin-contracts/contracts/token/ERC20/IERC20__factory";
export type { IXtokens } from "./lib/predeploy-contracts/contracts/xtokens/IXtokens";
export { IXtokens__factory } from "./factories/lib/predeploy-contracts/contracts/xtokens/IXtokens__factory";
export type { ERC20 } from "./lib/solmate/src/tokens/ERC20";
export { ERC20__factory } from "./factories/lib/solmate/src/tokens/ERC20__factory";
export type { ITokenBridge } from "./lib/wormhole/ethereum/contracts/bridge/interfaces/ITokenBridge";
export { ITokenBridge__factory } from "./factories/lib/wormhole/ethereum/contracts/bridge/interfaces/ITokenBridge__factory";
export type { IWETH } from "./lib/wormhole/ethereum/contracts/bridge/interfaces/IWETH";
export { IWETH__factory } from "./factories/lib/wormhole/ethereum/contracts/bridge/interfaces/IWETH__factory";
export type { IWormhole } from "./lib/wormhole/ethereum/contracts/interfaces/IWormhole";
export { IWormhole__factory } from "./factories/lib/wormhole/ethereum/contracts/interfaces/IWormhole__factory";
export type { BaseRouter } from "./src/BaseRouter";
export { BaseRouter__factory } from "./factories/src/BaseRouter__factory";
export type { Factory } from "./src/Factory";
export { Factory__factory } from "./factories/src/Factory__factory";
export type { FeeRegistry } from "./src/FeeRegistry";
export { FeeRegistry__factory } from "./factories/src/FeeRegistry__factory";
export type { MockToken } from "./src/MockToken";
export { MockToken__factory } from "./factories/src/MockToken__factory";
export type { WormholeRouter } from "./src/WormholeRouter";
export { WormholeRouter__factory } from "./factories/src/WormholeRouter__factory";
export type { XcmRouter } from "./src/XcmRouter";
export { XcmRouter__factory } from "./factories/src/XcmRouter__factory";
