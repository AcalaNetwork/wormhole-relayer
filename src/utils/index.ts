export * from './configureEnv';
export * from './logger';
export * from './relay';
export * from './route';
export * from './utils';
export * from './validate';
export * from './wormhole';
export * from './address';
export * from './error';
