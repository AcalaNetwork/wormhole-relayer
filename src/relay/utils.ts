import {
  hexToUint8Array,
  redeemOnEth,
  parseTransferPayload,
  importCoreWasm,
  ChainId,
  hexToNativeString,
} from '@certusone/wormhole-sdk';
import { ethers } from 'ethers';
import { ChainConfigInfo } from '../configureEnv';
import { EvmRpcProvider } from '@acala-network/eth-providers';
import {
  RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS,
} from './consts';
import axios from 'axios';
import { formatEther } from 'ethers/lib/utils';

interface VaaInfo {
  amount: bigint;
  originAddress: string;
  originChain: ChainId;
  targetAddress: string;
  targetChain: ChainId;
}
interface ShouldRelayResult {
  shouldRelay: boolean;
  msg: string;
} 

export const parseVaa = async (bytes: Uint8Array): Promise<VaaInfo> => {
  const { parse_vaa } = await importCoreWasm();
  const parsedVaa = parse_vaa(bytes);
  const buffered = Buffer.from(new Uint8Array(parsedVaa.payload));

  return parseTransferPayload(buffered);
};

export const shouldRelayVaa = (vaaInfo: VaaInfo): ShouldRelayResult => {
  const {
    amount,
    targetChain,
    originChain,
    originAddress,
  } = vaaInfo;
  const originAsset = hexToNativeString(originAddress, originChain);

  const res = shouldRelay({ targetChain, originAsset, amount });

  console.log('should relay: ', { targetChain, originAsset, amount, res });
  return res;
};

export const shouldRelay = ({
  targetChain,
  originAsset,
  amount: _amount,
}: {
  targetChain: number;
  originAsset: string;
  amount: bigint;
}): ShouldRelayResult => {
  const _noRelay = (msg: string): ShouldRelayResult => ({ shouldRelay: false, msg });

  if (!targetChain) return _noRelay('missing targetChain');
  if (!originAsset) return _noRelay('missing originAsset');
  if (!_amount) return _noRelay('missing transfer amount');

  let amount: bigint;
  try {
    amount = BigInt(_amount);
  } catch (e) {
    return _noRelay(`failed to parse amount: ${_amount}`);
  }

  const supported = RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS[targetChain];
  if (!supported) return _noRelay('target chain not supported');

  const minTransfer = supported[originAsset.toLowerCase()];
  if (!minTransfer) return _noRelay('token not supported');
  if (amount < BigInt(minTransfer)) return _noRelay(`transfer amount too small, expect at least ${minTransfer}`);

  return { shouldRelay: true, msg: '' };
};

export const relayEVM = async (
  chainConfigInfo: ChainConfigInfo,
  signedVAA: string,
  request: any,
  response: any
) => {
  const provider = EvmRpcProvider.from(chainConfigInfo.substrateNodeUrl);
  await provider.isReady();

  const signer = new ethers.Wallet(chainConfigInfo.walletPrivateKey, provider);

  const receipt = await redeemOnEth(
    chainConfigInfo.tokenBridgeAddress,
    signer,
    hexToUint8Array(signedVAA),
  );

  console.log('successfully redeemed on evm', receipt);
  response.status(200).json(receipt);
};

export const fetchBalance = async (address: string, url: string): Promise<number> => {
  try {
    const response = await axios.get(url, {
      data: {
        id: 0,
        jsonrpc: '2.0',
        method: 'eth_getBalance',
        params: [
          address,
          'latest',
        ],
      }
    });

    return Number(formatEther(response.data.result));
  } catch (e) {
    console.log('fetchBalance error: ', e);
    return -1;
  }
};
