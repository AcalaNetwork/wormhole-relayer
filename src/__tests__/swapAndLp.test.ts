import { ADDRESSES } from '@acala-network/asset-router/dist/consts';
import { BigNumber } from 'ethers';
import { ERC20__factory } from '@certusone/wormhole-sdk/lib/cjs/ethers-contracts';
import { FeeRegistry__factory } from '@acala-network/asset-router/dist/typechain-types';
import { describe, expect, it } from 'vitest';
import { formatUnits, parseEther, parseUnits } from 'ethers/lib/utils';
import { toHuman } from '@acala-network/asset-router/dist/utils';

import {
  JITOSOL_ADDR,
  JITOSOL_DECIMALS,
  TEST_ADDR_RELAYER,
  TEST_ADDR_USER,
} from './testConsts';
import { SwapAndLpParams } from '../utils';
import {
  api,
  expectError,
  provider,
  relayer,
  sudoTransferToken,
  transferToken,
  user,
} from './testUtils';

const recipient = TEST_ADDR_USER;
const poolId = '7';
const swapAmount = parseUnits('0.12', JITOSOL_DECIMALS).toString();
const minShareAmount = '0';

describe.concurrent('/shouldRouteSwapAndLp', () => {
  const testShouldRouteSwapAndLp = async (params: SwapAndLpParams) => {
    let res = await api.shouldRouteSwapAndLp(params);
    expect(res).toMatchSnapshot();

    // should be case insensitive
    res = await api.shouldRouteSwapAndLp({
      ...params,
      recipient: params.recipient.toLowerCase(),
    });
    expect(res).toMatchSnapshot();
  };

  it('when should route', async () => {
    await testShouldRouteSwapAndLp({
      recipient,
      poolId,
      swapAmount,
      minShareAmount,
    });
  });

  describe('when should not route', () => {
    it('when missing params', async () => {
      try {
        await api.shouldRouteSwapAndLp({
          recipient,
          poolId,
          minShareAmount,
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['swapAmount is a required field'], 400);
      }

      try {
        await api.shouldRouteSwapAndLp({
          recipient,
          swapAmount,
          minShareAmount,
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['poolId is a required field'], 400);
      }

      try {
        await api.shouldRouteSwapAndLp({
          poolId,
          swapAmount,
          minShareAmount,
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['recipient is a required field'], 400);
      }
    });

    it('when bad params', async () => {
      const res = await api.shouldRouteSwapAndLp({
        recipient,
        poolId: '999',
        swapAmount,
        minShareAmount,
      });
      expect(res).toMatchInlineSnapshot(`
        {
          "data": {
            "msg": "euphrates poolId 999 is not supported",
            "shouldRoute": false,
          },
        }
      `);
    });
  });
});

describe('route and rescue', () => {
  let routerAddr: string;
  const jitosol = ERC20__factory.connect(JITOSOL_ADDR, relayer);

  const fetchTokenBalances = async () => {
    if (!routerAddr) throw new Error('routerAddr not set');

    const [
      userBal,
      userTokenBal,
      relayerTokenBal,
      routerTokenBal,
    ] = await Promise.all([
      provider.getBalance(TEST_ADDR_USER),
      jitosol.balanceOf(TEST_ADDR_USER),
      jitosol.balanceOf(TEST_ADDR_RELAYER),
      jitosol.balanceOf(routerAddr),
    ]);

    console.log({
      userBal: toHuman(userBal, 18),
      userTokenBal: toHuman(userTokenBal, JITOSOL_DECIMALS),
      relayerTokenBal: toHuman(relayerTokenBal, JITOSOL_DECIMALS),
      routerTokenBal: toHuman(routerTokenBal, JITOSOL_DECIMALS),
    });

    return {
      userBal,
      userTokenBal,
      relayerTokenBal,
      routerTokenBal,
    };
  };

  it('/routeSwapAndLp', async () => {
    const relayerBal = await relayer.getBalance();
    expect(relayerBal.gt(parseEther('10'))).to.be.true;

    const routeArgs = {
      recipient: user.address,
      poolId,
      swapAmount,
      minShareAmount,
    };

    const shouldRouteRes = await api.shouldRouteSwapAndLp(routeArgs);
    ({ routerAddr } = shouldRouteRes.data);
    console.log({ routerAddr });

    // make sure user has enough token to transfer to router
    const bal = await fetchTokenBalances();
    const trasnferAmount = BigNumber.from(swapAmount).mul(2);
    if (bal.userTokenBal.lt(trasnferAmount)) {
      if (bal.relayerTokenBal.lt(trasnferAmount)) {
        throw new Error('both relayer and user do not have enough jitosol to transfer to router!');
      }

      console.log('refilling token for user ...');
      await (await jitosol.transfer(TEST_ADDR_USER, trasnferAmount)).wait();
    }

    console.log('transferring token to router ...');
    const transferAmountHuman = Number(formatUnits(trasnferAmount, JITOSOL_DECIMALS));
    await transferToken(routerAddr, user, JITOSOL_ADDR, transferAmountHuman);

    const bal0 = await fetchTokenBalances();

    console.log('routing ...');
    const routeRes = await api.routeSwapAndLp({
      ...routeArgs,
      token: JITOSOL_ADDR,
    });
    const { txHash, removed } = routeRes.data;
    console.log(`route finished! txHash: ${txHash}, removed: ${removed}`);

    const bal1 = await fetchTokenBalances();

    // router should be destroyed
    const routerCode = await provider.getCode(routerAddr);
    expect(routerCode).to.eq('0x');
    expect(bal1.routerTokenBal.toNumber()).to.eq(0);

    // relayer should receive routing fee and swap fee
    const routingFee = await FeeRegistry__factory.connect(ADDRESSES.ACALA.feeAddr, provider)
      .getFee(JITOSOL_ADDR);
    const swapFee = parseUnits('0.0035', 9);
    expect(bal1.relayerTokenBal.sub(bal0.relayerTokenBal).toBigInt())
      .to.eq(routingFee.add(swapFee).toBigInt());

    // user should receive 3 ACA drop
    expect(bal1.userBal.sub(bal0.userBal).toBigInt()).to.eq(parseEther('3').toBigInt());
  });

  it('/rescueSwapAndLp', async () => {
    const relayerBal = await relayer.getBalance();
    expect(relayerBal.gt(parseEther('10'))).to.be.true;

    const routeArgs = {
      recipient: user.address,
      poolId,
      swapAmount,
      minShareAmount,
    };

    const shouldRouteRes = await api.shouldRouteSwapAndLp(routeArgs);
    ({ routerAddr } = shouldRouteRes.data);
    console.log({ routerAddr });

    // make sure user has enough token to transfer to router
    const bal = await fetchTokenBalances();
    const trasnferAmount = BigNumber.from(swapAmount).mul(2);
    if (bal.userTokenBal.lt(trasnferAmount)) {
      if (bal.relayerTokenBal.lt(trasnferAmount)) {
        throw new Error('both relayer and user do not have enough jitosol to transfer to router!');
      }

      console.log('refilling token for user ...');
      await (await jitosol.transfer(TEST_ADDR_USER, trasnferAmount)).wait();
    }

    console.log('transferring token to router ...');
    const transferAmountHuman = Number(formatUnits(trasnferAmount, JITOSOL_DECIMALS));
    await transferToken(routerAddr, user, JITOSOL_ADDR, transferAmountHuman);

    const bal0 = await fetchTokenBalances();

    console.log('rescuing ...');
    const rescueRes = await api.rescueSwapAndLp({
      ...routeArgs,
      token: JITOSOL_ADDR,
    });
    const txHash = rescueRes.data;
    console.log(`rescue finished! txHash: ${txHash}`);

    const bal1 = await fetchTokenBalances();

    // router should be destroyed
    const routerCode = await provider.getCode(routerAddr);
    expect(routerCode).to.eq('0x');
    expect(bal1.routerTokenBal.toNumber()).to.eq(0);

    // relayer should receive swap fee
    const swapFee = parseUnits('0.0035', 9);
    expect(bal1.relayerTokenBal.sub(bal0.relayerTokenBal).toBigInt()).to.eq(swapFee.toBigInt());

    // user should receive 3 ACA drop and token back
    expect(bal1.userBal.sub(bal0.userBal).toBigInt()).to.eq(parseEther('3').toBigInt());
    expect(bal1.userTokenBal.sub(bal0.userTokenBal).toBigInt())
      .to.eq(trasnferAmount.sub(swapFee).toBigInt());
  });
});


