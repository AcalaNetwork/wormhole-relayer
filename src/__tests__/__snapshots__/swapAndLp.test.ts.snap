// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`/shouldRouteSwapAndLp > when should route 1`] = `
{
  "data": {
    "routerAddr": "0x8ed211e6e940955A8ccF2ec2BA792eDfd7Cca9bD",
    "shouldRoute": true,
  },
}
`;

exports[`/shouldRouteSwapAndLp > when should route 2`] = `
{
  "data": {
    "routerAddr": "0x8ed211e6e940955A8ccF2ec2BA792eDfd7Cca9bD",
    "shouldRoute": true,
  },
}
`;
